tensorpipe/transport/shm/connection.cc

```diff
@@ -358,6 +362,14 @@ Connection::WriteOperation::WriteOperation(
     : ptr_(ptr), len_(len), fn_(std::move(fn)) {}
 
 void Connection::WriteOperation::handleWrite(TProducer& outbox) {
+  // Attempting to write a message larger than the ring buffer. We might want to
+  // chunk it in the future.
+  const int buf_size = outbox.getHeader().kDataPoolByteSize;
+  if (len_ > buf_size) {
+    fn_(TP_CREATE_ERROR(ShortWriteError, len_, buf_size));
+    return;
+  }
```

Comment:
What happens without this check? Is there a check in the ringbuffer itself that prevents this from happening?

Reply:
Without this check, the `copyToRingBuffer_` method fails [here](https://github.com/pytorch/tensorpipe/blob/master/tensorpipe/util/ringbuffer/producer.h#L201), which makes `writeInTxWithSize` fail [here](https://github.com/pytorch/tensorpipe/blob/master/tensorpipe/util/ringbuffer/producer.h#L90), which makes `handleWrite` throw `system_error` [here](https://github.com/pytorch/tensorpipe/blob/master/tensorpipe/transport/shm/connection.cc#L376).

We could check the return value of `writeInTxWithSize` but it will return `EINVAL` in different cases (write of size zero, write of size larger than buffer size, write outside of a transaction), hence the specific check here. Tehcnically, a write outside transaction should not happen unless we mess up, and a zero-write can easily be identified, so we could indeed use `EINVAL` to mean "write larger than buffer size" (if the buffer is too full, the error returned is `ENOSPC`, so we can distinguish the two).